File: solutions/0206.go
package solutions

import (
	"reflect"

	"pokman/bulbasaur/leetcode/ds"
)

func wrapReverseList(a []int) []int {
	return ds.ListToArr(reverseList(ds.MakeListNode(a)))
}

func reverseList(head *ds.ListNode) *ds.ListNode {
	if head.Next == nil {
		return head
	}

	/* recursive
	ret := reverseList(head.Next)

	if head.Next.Next == nil {
		head.Next.Next = head
		head.Next = nil
	}
	*/
	var pre *ds.ListNode = nil
	var cur *ds.ListNode = head
	var nxt *ds.ListNode = head.Next
	for nxt != nil {
		new_cur := nxt
		nxt = nxt.Next
		cur.Next = pre
		pre = cur
		cur = new_cur
	}
	cur.Next = pre

	return cur
}

func init() {
	desc := `
Reverse a singly linked list.

Example:
Input: 1->2->3->4->5->NULL
Output: 5->4->3->2->1->NULL

Follow up:
A linked list can be reversed either iteratively or recursively. Could you implement both?
	`
	sol := Solution{
		Title:  "Reverse Linked List",
		Desc:   desc,
		Method: reflect.ValueOf(wrapReverseList),
		Tests:  make([]TestCase, 0),
	}
	a := TestCase{}
	a.Input = []interface{}{[]int{1, 2, 3, 4, 5}}
	a.Output = []interface{}{[]int{5, 4, 3, 2, 1}}
	sol.Tests = append(sol.Tests, a)

	SolutionMap["0206"] = sol
}

File: solutions/0206.rs
// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    #[allow(dead_code)]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

struct Solution {}

impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        if head.is_none() {
            return head;
        }

        match head.as_ref() {
            Some(node) => {
                if node.next.is_none() {
                    return head;
                }
            }
            None => {}
        }

        let mut pre = None;
        let mut cur = head;

        while let Some(mut node) = cur {
            let next = node.next;
            node.next = pre;
            pre = Some(node);
            cur = next;
        }

        pre
    }
}

fn main() {
    println!("problem 206:");
    println!("-------------------------------");

    let mut input = Some(Box::new(ListNode::new(1)));
    match input.as_mut() {
        Some(mut node) => {
            node.next = Some(Box::new(ListNode::new(2)));
        }
        None => {}
    }

    println!("input: {:?}", input);

    let mut output = Solution::reverse_list(input);
    println!("output: {:?}", output);
    while let Some(node) = output {
        println!("node: {}", node.val);
        output = node.next;
    }
}
